### PR TITLE
Aftershock: Port Augustmoon Station

### PR DESCRIPTION
**Pull request by John-Candlebury**
#### Summary
Mods "Aftershock: Add the Port Augustmoon Station"

#### Purpose of change
For a long time Aftershock has wanted an NPC Hub area that is both safe from the terrible cold and lack of food and allows for easier trade and turning in or obtaining extra missions to complete during expeditions.

The Port Augustmoon Station is intended to be this space.

#### Describe the solution

This is a space station you can move yourself your cargo and your companion to/from through the magic of EOCs. This is triggered by either calling a drone shuttle (expensive) or approaching a manned surface landing pad (cheaper but I mean its still space travel).  

The player is intended to spawn here during a regular non-challenge aftershock mod start and to return here after most succesfull  expeditions to the planet below, so they can actually trade salvage buy advanced technology/supplies and obtain reconnaissance missions to perform planet side.

Heres how the station looks like atm (likely non final)
image

List of necessary features (probably will expand a bit yet)

- [x] Finished Station Mapgen
- [x] NPC traders and quest givers similar to those in the Ancilla Bar.
- [x] Surface landing pad mapgen and related EOCs and NPCs.
- [x] Shuttle to the station.
- [x] More station unique json content (non critical) 

#### Describe alternatives you've considered

Putting this thing in the ground, but that makes it easier for the player to destroy or exploit (not that this is currently hard)

#### Testing
WIP the tests will error out.

### CODE REVIEW

File: data/mods/Aftershock/maps/mapgen/port_augustmoon/port_augustmoon_main.json [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+        "                        ",
+        "                        ",
+        "                        ",
+        "                        ",
+        "                        ",
+        "                        "
+      ],
+      "palettes": [ "afs_habitat_structure", "afs_habitat_residential_furnishing" ],
+      "terrain": { "1": "t_dirtmound", " ": "t_hard_vacuum", "2": "t_bars", "5": "t_ramp_down_high", "6": "t_ramp_down_low" },
+      "furniture": { "p": "f_solar_unit" }
```
Comment (ehughsbaird) [inline data/mods/Aftershock/maps/mapgen/port_augustmoon/port_augustmoon_main.json:153]: Spacecraft typically have radiator arrays also, due to the difficulty of losing heat in space. I don't know if the mod setting changes that or you care about that for this.

Comment (John-Candlebury) [reply data/mods/Aftershock/maps/mapgen/port_augustmoon/port_augustmoon_main.json:153]: I do but probably wont care at this specific point in time haha. Its not like the station systems actually do anything.

**Comment by John-Candlebury** [comment]
The hull terrains are indestructible (or at least intended to be so I gotta check). 

Although In honest I don't care about people sabotaging themselves like that.

**Comment by MNG-cataclysm** [comment]
I did some checking, and it turns out that the Port Augustmoon station terrain is actually quite destructible, 50L barrel bombs are sufficient to destroy it.

File: data/mods/Aftershock/maps/furniture_and_terrain/terrain_spaceship.json [hunk 2]
```diff
@@ -22,6 +22,29 @@
...
+    "flags": [ "TRANSPARENT", "NOITEM", "WALL", "NO_SCENT", "AUTO_WALL_SYMBOL", "OPENCLOSE_INSIDE", "BLOCK_WIND" ]
+  },
+  {
+    "type": "terrain",
+    "id": "t_afs_augustmoon_hull_wall",
+    "name": "Port Augustmoon hull",
+    "description": "The ancient hull of Port Augustmoon, spun into orbit several lifetimes ago.",
+    "roof": "t_metal_flat_roof",
+    "copy-from": "t_wall_metal",
+    "flags": [ "NOITEM", "SUPPORTS_ROOF", "WALL", "NO_SCENT", "AUTO_WALL_SYMBOL", "MINEABLE", "BLOCK_WIND" ],
```
Comment (MNG-cataclysm) [inline data/mods/Aftershock/maps/furniture_and_terrain/terrain_spaceship.json:44]:
```suggestion
    "flags": [ "NOITEM", "SUPPORTS_ROOF", "WALL", "NO_SCENT", "AUTO_WALL_SYMBOL",  "BLOCK_WIND" ],
```
Players probably shouldn't be able to mine these.

**Review by MNG-cataclysm — CHANGES_REQUESTED** [review]

File: data/mods/Aftershock/maps/furniture_and_terrain/terrain_spaceship.json [hunk 3]
```diff
@@ -22,6 +22,29 @@
...
+  {
+    "type": "terrain",
+    "id": "t_afs_augustmoon_diamond_glass",
+    "name": "diamond-matrix glass",
+    "description": "The diamond cubic lattice distorts distant light sources into eight pointed stars.  The galactic core is visible through the glass, but no spacer would ever give it more than a passing glance.",
+    "roof": "t_metal_flat_roof",
+    "symbol": "LINE_XOXO",
+    "color": "light_cyan",
+    "move_cost": 0,
+    "coverage": 100,
```
Comment (MNG-cataclysm) [inline data/mods/Aftershock/maps/furniture_and_terrain/terrain_spaceship.json:34]:
```suggestion
    "coverage": 100,
    "bash": {
      "str_min": 4000,
      "str_max": 8000,
      "sound": "whump!",
      "sound_fail": "whump!",
      "ter_set": "t_afs_augustmoon_diamond_glass"
    },
```
If you would prefer the station terrain itself to be indestructible for the time being, I've found that having a terrain break into itself accomplishes this.

File: data/mods/Aftershock/maps/furniture_and_terrain/terrain_spaceship.json [hunk 4]
```diff
@@ -22,6 +22,29 @@
...
+    "flags": [ "TRANSPARENT", "NOITEM", "WALL", "NO_SCENT", "AUTO_WALL_SYMBOL", "OPENCLOSE_INSIDE", "BLOCK_WIND" ]
+  },
+  {
+    "type": "terrain",
+    "id": "t_afs_augustmoon_hull_wall",
+    "name": "Port Augustmoon hull",
+    "description": "The ancient hull of Port Augustmoon, spun into orbit several lifetimes ago.",
+    "roof": "t_metal_flat_roof",
+    "copy-from": "t_wall_metal",
```
Comment (MNG-cataclysm) [inline data/mods/Aftershock/maps/furniture_and_terrain/terrain_spaceship.json:43]:
```suggestion
    "copy-from": "t_wall_metal",
    "bash": { "str_min": 4000, "str_max": 8000, "sound": "clang!", "sound_fail": "clang!", "ter_set": "t_afs_augustmoon_hull_wall" },
```
Same as before with the diamond glass.

**Comment by MNG-cataclysm** [comment]
Why not add the redesigned NPC factions and their NPCs to the Exoplanet mod folder itself, rather than the main Aftershock folder? That keeps old interactions intact if players manually disable Exoplanet, and allows for a more self-contained categorization of Exoplanet content; you don't need to sift through non-Exoplanet stuff to find what you're looking for.

**Comment by John-Candlebury** [comment]
All Aftershock content is developed with the intention that it belongs in the exoplanet's setting, both mods simply arent meant to be distinct from each other, they are just split so people arent forced to play in the unfinished total conversion.

To simplify merging both mods in the future and to ensure features get a better playtesting, all mod content that isn't directly involved in transforming the play area into an icy planet is added to the main Aftershock mod.

**Comment by kasanryukin** [comment]
As mentioned on discord, there is some wonkiness with getting a shuttle to land near you, and the mission to do so is absent.  A second beacon didn't spawn a new site, and a new world failed to even spawn the first site.   I did find a space port and got TO the station, so that part works as intended.

**Comment by John-Candlebury** [comment]
I'd appreciate a more detailed reproduction steps for the shuttle issue if you can.
Tried a beacon once to have a shuttle spawn succesfully, allowed it to leave by waiting 6 hours, and then succesfully called two others, once while away from the previous crater and once while standing over it. It all worked properly.

**Comment by kasanryukin** [comment]
> I'd appreciate a more detailed reproduction steps for the shuttle issue if you can. Tried a beacon once to have a shuttle spawn succesfully, allowed it to leave by waiting 6 hours, and then succesfully called two others, once while away from the previous crater and once while standing over it. It all worked properly.

In the first case, I spawned into a new world on a random character, being assigned the scenario where you have to refuel your helicopter.  I spawned the drone, activated it, and got the popup message that it would land nearby.  No mission popped up, and I didn't know that it can spawn outside the viewable overmap, so I waited 12 hours, before you told me what to look for on discord.  I found the site where it should have landed and it was just a pile of upturned earth.   I spawned a second beacon and used it, but not new shuttle landing spot appear with in the current, or three nearest overmap regions (in case revealing the region prevented a spawn in said region.)

Operating under the assumption I shouldn't have revealed the map, I reset the world, got a new random character and spawned a new beacon. Again I got the popup message, and this time I just climb to the roof of the evac shelter to expose more overmap.  I was unable to find the landing sight after around 2 in game hours, and since you said it despawns after for, I went ahead and revealed the overmap.  No landing site was generated in that world.  There was a space port in the far NE corner of the region, so I just teleported over it to investigate it. After seeing the staggering cost and spawning enough merch to pay for it, I accepted the teleport and was successfully move to Port Augustmoon, which had spawned at Z -10 near the southern border of my current region. (It spawned a mere 4 overmap tiles away from the secret at the bottom of a physics lab.)

I didn't conduct a third world test yet.

**Comment by Maleclypse** [comment]
The two best commits this year 
```
* Unending Toil

* Eternal toil amongst the stars
```

**Comment by John-Candlebury** [comment]
Woo! It's in means I gotta expand it now haha.
Thanks Mal!

**Pull request merged**